crates/admin/src/cluster_controller/service.rs

```diff
@@ -243,114 +243,148 @@ impl<T: TransportConnect> Service<T> {
-            configuration,
+        let logs_controller = LogsController::init(
+            &configuration,
             self.metadata.clone(),
             self.bifrost.clone(),
             self.metadata_store_client.clone(),
             self.metadata_writer.clone(),
         )
         .await?;
 
-        let mut observed_cluster_state = ObservedClusterState::default();
+        let (log_trim_interval, log_trim_threshold) =
+            create_log_trim_interval(&configuration.admin);
 
-        let mut logs_watcher = self.metadata.watch(MetadataKind::Logs);
-        let mut partition_table_watcher = self.metadata.watch(MetadataKind::PartitionTable);
-        let mut logs = self.metadata.updateable_logs_metadata();
-        let mut partition_table = self.metadata.updateable_partition_table();
-        let mut nodes_config = self.metadata.updateable_nodes_config();
         let mut find_logs_tail_interval =
             time::interval(configuration.admin.log_tail_update_interval.into());
         find_logs_tail_interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
 
+        let leader = Leader {
+            metadata: self.metadata.clone(),
+            bifrost: self.bifrost.clone(),
+            metadata_store_client: self.metadata_store_client.clone(),
+            metadata_writer: self.metadata_writer.clone(),
+            logs_watcher: self.metadata.watch(MetadataKind::Logs),
+            nodes_config: self.metadata.updateable_nodes_config(),
+            partition_table: self.metadata.updateable_partition_table(),
+            partition_table_watcher: self.metadata.watch(MetadataKind::PartitionTable),
+            cluster_state_watcher: self.cluster_state_refresher.cluster_state_watcher(),
+            find_logs_tail_interval,
+            log_trim_interval,
+            log_trim_threshold,
+            logs: self.metadata.updateable_logs_metadata(),
+            logs_controller,
+            scheduler,
+        };
+
+        Ok(leader)
+    }
+
+    pub async fn run(mut self) -> anyhow::Result<()> {
+        self.init_partition_table().await?;
+
+        let mut config_watcher = Configuration::watcher();
+        let mut cluster_state_watcher = self.cluster_state_refresher.cluster_state_watcher();
+
+        self.task_center.spawn_child(
+            TaskKind::SystemService,
+            "cluster-controller-metadata-sync",
+            None,
+            sync_cluster_controller_metadata(self.metadata.clone()),
+        )?;
+
+        let mut shutdown = std::pin::pin!(cancellation_watcher());
+
+        let bifrost_admin = BifrostAdmin::new(
+            &self.bifrost,
+            &self.metadata_writer,
+            &self.metadata_store_client,
+        );
+
+        let mut state: ClusterControllerState<T> = ClusterControllerState::Follower;
+
         self.health_status.update(AdminStatus::Ready);
 
         loop {
+            self.next_cluster_state(&mut state).await?;
```

Comment:
I think this can be happen on cluster_state change only since (imho) it is the only way the leadership state can change, correct?

Reply:
Yes, I think so.